storage/src/main/java/org/apache/kafka/storage/internals/log/LogSegment.java

```diff
@@ -0,0 +1,873 @@
+        offsetIndex.resize(offsetIndex.maxIndexSize());
+        timeIndex.resize(timeIndex.maxIndexSize());
+
+        int bytesTruncated;
+        if (mapping == null)
+            bytesTruncated = 0;
+        else
+            bytesTruncated = log.truncateTo(mapping.position);
+
+        if (log.sizeInBytes() == 0) {
+            created = time.milliseconds();
+            rollingBasedTimestamp = OptionalLong.empty();
+        }
+
+        bytesSinceLastIndexEntry = 0;
+        if (maxTimestampSoFar() >= 0)
+            maxTimestampAndOffsetSoFar = readLargestTimestamp();
+
+        return bytesTruncated;
+    }
+
+    private TimestampOffset readLargestTimestamp() throws IOException {
+        // Get the last time index entry. If the time index is empty, it will return (-1, baseOffset)
+        TimestampOffset lastTimeIndexEntry = timeIndex().lastEntry();
+        OffsetPosition offsetPosition = offsetIndex().lookup(lastTimeIndexEntry.offset);
+
+        // Scan the rest of the messages to see if there is a larger timestamp after the last time index entry.
+        FileRecords.TimestampAndOffset maxTimestampOffsetAfterLastEntry = log.largestTimestampAfter(offsetPosition.position);
+        if (maxTimestampOffsetAfterLastEntry.timestamp > lastTimeIndexEntry.timestamp)
+            return new TimestampOffset(maxTimestampOffsetAfterLastEntry.timestamp, maxTimestampOffsetAfterLastEntry.offset);
+
+        return lastTimeIndexEntry;
+    }
+
+    /**
+     * Calculate the offset that would be used for the next message to be append to this segment.
+     * Note that this is expensive.
+     *
+     * This method is thread-safe.
+     */
+    public long readNextOffset() throws IOException {
+        FetchDataInfo fetchData = read(offsetIndex().lastOffset(), log.sizeInBytes());
+        if (fetchData == null)
+            return baseOffset;
+        else
+            return fetchData.records.lastBatch()
+                .map(batch -> batch.nextOffset())
+                .orElse(baseOffset);
+    }
+
+    /**
+     * Flush this log segment to disk.
+     *
+     * This method is thread-safe.
+     */
+    public void flush() throws IOException {
+        try {
+            LOG_FLUSH_TIMER.time(new Callable<Void>() {
+                // lambdas cannot declare a more specific exception type, so we use an anonymous inner class
+                @Override
+                public Void call() throws IOException {
+                    log.flush();
+                    offsetIndex().flush();
+                    timeIndex().flush();
+                    txnIndex.flush();
+                    return null;
+                }
+            });
+        } catch (Exception e) {
+            if (e instanceof IOException)
+                throw (IOException) e;
+            else if (e instanceof RuntimeException)
+                throw (RuntimeException) e;
+            else
+                throw new IllegalStateException("Unexpected exception thrown: " + e, e);
```

Comment:
1. This is a new change, right? Can you please explain how IllegalState is handled?

2. Do we need to print the entire stack trace in the exception message? Perhaps simply "Unexpected exception thrown" should suffice (since we are already passing the entire exception)

Reply:
Not really, this is a change to make the compiler happy, but we never throw a checked exception that is not an IOException. If the code ever changes to do that, we'll catch it here. I think it's fine as it is.

Reply:
ok